Re-add pq-devnet-0.md for team collaboration (#4)
leanEthereum/pm: pq-interop/pq-devnet-0.md

# PQ Devnet-0 Specs

### Devnet Change Log
- No changes, initial devnet 

### Active Discussions

> - **Sig Parameters:** Consensus on signature parameters (64 hash chains, chain length 8, max lifetime 2^18)? 

Reviewer: @b-wagn Apologies if this sounds naive, but shouldn’t the hash-sig crate expose the underlying Merkle tree [here](https://github.com/b-wagn/hash-sig/blob/2e455dbfd89057640c93795c47f1bf1e62b0d461/src/signature/generalized_xmss.rs#L162)?
That way, beam clients could iterate over the tree and use each leaf for signing.

> - **Sig Parameters:** Consensus on signature parameters (64 hash chains, chain length 8, max lifetime 2^18)? 

Reviewer 2: +1

- **Slot Duration:** 4s (Zeam) vs. 12s (Ream) – agree on one? 

> - **Networking Transport:** QUIC (Ream) vs. TCP first then QUIC (discussion in meeting)? 

Reviewer 3: Because there is a lot of work to do for support Gossipsub v1.2, zig-libp2p will implement v1.0 first. However through multistream select, I think we don't need limit the Gossipsub version. It will detect the compatible version automatic.

> - **Networking Transport:** QUIC (Ream) vs. TCP first then QUIC (discussion in meeting)? 

Reviewer 4: Can we mandate dual-stack support clients must handle both QUIC and TCP.

> - **Networking Transport:** QUIC (Ream) vs. TCP first then QUIC (discussion in meeting)? 

Reviewer: > Can we mandate dual-stack support clients must handle both QUIC and TCP.

The objective is for all clients to migrate to QUIC in PQ world. Almost all clients already support QUIC. I think the challenge right now is the availability of QUIC support in zig libp2p libraries.

- **Validator Count:** Start with 1,000 keys (meeting consensus) and scale? 

> - **Consensus Mechanism:** Include minimal 3SF or no global consensus? 

Reviewer 5: This was a great idea by @KolbyML to help developers keep making progress even if aggregation is not ready until after the first iteration. Changing the order would pipeline tasks.

> - **Consensus Mechanism:** Include minimal 3SF or no global consensus? 

Reviewer 6: From meeting:

0. chaining + minimal fork choice
1. pqsig
2. single subnet aggregation

- **Infrastructure Setup:** Who handles (e.g., Grafana, Prometheus, Kurtosis-like for interop, Genesis Generator tool )? 

## Sub-Spec List for PQ Devnet-0
>The list below links to reference papers, specs, and implementations. Versions to be finalized by August 31, 2025.

### Test Releases

**Consensus Specs:** link to consensus spec release
**Execution Spec Tests:** link to executuon spec release 

### Spec Versions Required & Open PRs

- List of links to spec PR's **Merged** :heavy_check_mark: 
- List of links to spec PR's **Open** :exclamation:

### Client Support Status
>develop list of testing scenarios (link here)

| Testing Scenario | Ream | Zeam | [TBD: Third Client?] |
|------------------|------|------|----------------------|
| 1a | :question: | :question: | :question: |
| 2a | :question: | :question: | :question: |
| 3a | :question: | :question: | :question: |
| 4a | :question: | :question: | :question: |
| 5a | :question: | :question: | :question: |
| 6a | :question: | :question: | :question: |

> ## Kurtosis Interop Config (Pre-Devnet Testing)

Reviewer 2: do we have anyone working to fork kurtosis to support the lean ethereum?

Author: picked up comment in 'active discussions' section

>yaml below is AI slop

```yaml
participants_matrix:

>   cl:

Author: picked up comment in 'active discussions' section

    - cl_type: ream
      cl_image: reamlabs/ream:pq-devnet-0
    - cl_type: zeam
      cl_image: zeam/zeam:pq-devnet-0
    - cl_type: [TBD - third client, e.g., lighthouse-pq]
      cl_image: [TBD]:pq-devnet-0

network_params:

>   pq_fork_epoch: 1  # Activate PQ features at epoch 1

Reviewer 6: ```suggestion
```

[Personal view] We probably don't need fork epoch for the foreseeable future, but on a related note I believe @KolbyML has some thoughts around defining a fork digest.

Reviewer 5: If we have 3SF, there will be no epoch.

  max_validators: 1000  # Initial keys, scale to 10k+
  slot_time_seconds: [TBD - 4 or 12]

>   preset: lean  # Minimal lean consensus preset

Reviewer 2: beacon runs on two presets, so specifying that is what is intended here. we can retain lean minimal preset placeholder alth only implement one preset for now geared towards testing this devnet

>   preset: lean  # Minimal lean consensus preset

Reviewer 6: sgtm we can resolve this unchanged

fwiw it should eventually link to something like this: https://github.com/ethereum/consensus-spec-tests/blob/master/presets/minimal/phase0.yaml

additional_services:
  - metrics_dashboard  # Shared Prometheus/Grafana
  - genesis_generator

metrics_params:
  image: [TBD - ethpandaops/metrics:master]
  dashboards:
    - scenario: pq_signing
      config:
        throughput: 100  # Signatures per slot
    - scenario: propagation
      config:
        throughput: 50  # Blocks/attestations
```

### Testing:

>testing ideas below is AI slop

#### PQ Signature Viability

>   * Key generation and signing/verification timings - kurtosis - generate keys for 1,000 validators and measure sub-second targets.

Author: picked up comment in 'active discussions' section

  * Adversarial testing: Inject bad signatures via custom actors; verify rejection.
  * Aggregation prep: Optional local aggregation without network publish; measure times.
  * EIP-7870 compliance: Monitor CPU/memory/bandwidth during runs.

#### Multi-Client Interop
  * Sync testing: Start nodes at different points; use block-by-root to catch up.
  * Attestation correctness: Assertoor-like test for ≥98% (Zeam) or local tally (Ream).
  * Fork choice: Test with mini 3SF or no consensus; verify chain views align.
  * Networking: Gossip propagation delays <1s; QUIC/TCP comparison.

> 

Author: revised to address these comments

#### Resource Requirements

>   * Run with 1,000 validators; scale to 10,000; measure EIP-7870 limits.

Reviewer 2: ahh yes, 10X scale implies each new devnet iteration

>   * Run with 1,000 validators; scale to 10,000; measure EIP-7870 limits.

Author: revised to address these comments

  * Hardware metrics: CPU load, memory, throughput every second.
  * Bonus: Inject adversaries (signature collisions, censoring); monitor resilience.

#### Success Criteria Validation

>   * Network runs for 300+ slots * num_clients without crashing.

Reviewer 2: 300 slots we can  test and run locaaly, so yes if we are doing a cluster devnet, we should run it for longer. also will be useful when we add 3SF to next devnet iteration if we don't already have it in the 0rth

# PQ Devnet-0 Specs

### Devnet Change Log
- No changes, initial devnet 

### Active Discussions

- **Sig Parameters:**
    - Hash chains: 64
    - Chain length: 8
    - Max lifetime: 2^32
    - Activation time: 2^18
- **Slot Duration:** 4 seconds

> - **Networking Transport:** QUIC (Ream) vs. TCP first then QUIC (discussion in meeting)? 

Reviewer 6: ```suggestion
- **Networking Transport:** QUIC
```

- **Gossip:** Gossipsub v1.2

> - **Validator Count:** Start with 1,000 keys (meeting consensus) and scale? 

Reviewer 6: ```suggestion
- **Validator Count:** 1,000 keys (meeting consensus)
```

Reviewer 4: Hello @kamilsa I wrote out this testing plan, specifically because we have a small scale plan, which is perfect for this kind of test cases https://hackmd.io/@bomanaps/ryUB8Sywlx
would love to get your thoughts on this

Reviewer 7: @bomanaps Hey! Looks interesting. I believe with Kurtosis we should be able to implement most of these tests. Some of these tests we can actually implement in our signature aggregation [simulator](https://github.com/qdrvm/beamsim/)

> - **Consensus Mechanism:** Include minimal 3SF or no global consensus? 

Reviewer 6: ```suggestion
- **Consensus Mechanism:** Minimal fork choice rule (to be spec'ed)
```

- **Infrastructure Setup:** Should we engage ethPandaOps to:
    - Set up and host a shared Grafana and Prometheus instance for this effort?
    - Fork Kurtosis to support leanEthereum devnets (and patch in Genesis Generator)?
    - Use ethPandaOps Docker image builder and tracker to build/publish CL images?

## Interop Objectives, by Devnet
| Devnet-# | Objective(s) | Aspirational Ship Date |
| --------- | ------------ | ----------------------- |

> | **`pq-interop-0`** |  - Signature chaining<br>-Minimal fork choice | End of August |

Reviewer 6: ```suggestion
| **`pq-devnet-0`** |  - Signature chaining<br>-Minimal fork choice | End of August |
```

> | **`pq-interop-1`** | - PQ-Signatures | Mid-October |

Reviewer 6: ```suggestion
| **`pq-devnet-1`** | - PQ-Signatures | Mid-October |
```

> | **`pq-interop-2`** | - Single Subnet Aggregation | Mid-November |

Reviewer 6: ```suggestion
| **`pq-devnet-2`** | - Single Subnet Aggregation | TBD |
```

We still can't estimate the timeline for aggregation specs

    - cl_type: ream
      cl_image: reamlabs/ream:pq-devnet-0
    - cl_type: zeam
      cl_image: zeam/zeam:pq-devnet-0
    - cl_type: [TBD - third client, e.g., lighthouse-pq]
      cl_image: [TBD]:pq-devnet-0

network_params:
  max_validators: 1000  # Initial keys, scale to 10k+
  slot_time_seconds: [TBD - 4 or 12]

additional_services:
  - metrics_dashboard  # Shared Prometheus/Grafana
  - genesis_generator

metrics_params:
  image: [TBD - ethpandaops/metrics:master]
  dashboards:
    - scenario: pq_signing
      config:
        throughput: 100  # Signatures per slot
    - scenario: propagation
      config:
        throughput: 50  # Blocks/attestations
```

### Testing:

>testing ideas below is AI slop

#### PQ Signature Viability
  * Key generation and signing/verification timings - kurtosis - generate keys for 1,000 validators and measure sub-second targets.
  * Adversarial testing: Inject bad signatures via custom actors; verify rejection.
  * EIP-7870 compliance: Monitor CPU/memory/bandwidth during runs.

#### Multi-Client Interop
  * Block-by-root: Use for recovery to catch-up to chain head when nodes break; include chianing to pull parent chains for latest gossip blocks.

>   * Attestation correctness: Assertoor-like test for ≥98% (Zeam) or local tally (Ream).

Reviewer 6: ```suggestion
  * Attestation correctness: ≥98%
```

#### Resource Requirements
  * Run with 1,000 validators (devnet-0); scale by 10x validators for each iteration to 10,000; measure EIP-7870 limits.
  * Hardware metrics: CPU load, memory, throughput every second.
  * Bonus: Inject adversaries (signature collisions, censoring); monitor resilience.

#### Success Criteria Validation
  * Network runs for at least 300 slots * num_clients without crashing for local testing; for cluster devnet, aim for longer durations (e.g., 10 days), subject to test infrastructure capabilities (Extended runs will be beneficial for incorporating features like 3SF in subsequent devnet iterations).

>   * Full block pipeline <6s (Ream) or on-time with 4s slots (Zeam).

Reviewer 6: ```suggestion
  * Full block pipeline within <4 seconds
```

Reviewer 5: Maybe just `<4 seconds`? or `within 4 seconds`?

Reviewer 6: updated to in <4 seconds since it gives more clarity than "within"

>   * `num_attestations_received / num_validators  >= 0.098` for every slot (i.e. 98% attestation arrival)

Reviewer 6: ```suggestion
```

Stick to 98% attestation correctness for now

>   * `num_valid_signatures == num_attestations_received` for every slot (i.e. 100% attestation correctness)

Reviewer 6: ```suggestion
```

Stick to 98% attestation correctness for now
